provider/cldr/src/transform/datetime/mod.rs

```diff
@@ -2,8 +2,318 @@
 // called LICENSE at the top level of the ICU4X source tree
 // (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
 
-mod common;
-pub mod patterns;
-pub mod skeletons;
-pub mod symbols;
+use crate::cldr_serde;
+use crate::error::Error;
+use crate::reader::{get_langid_subdirectories, get_langid_subdirectory, open_reader};
+use crate::CldrPaths;
+use elsa::sync::FrozenBTreeMap;
+
+use icu_datetime::provider::calendar::*;
+use icu_provider::iter::IterableResourceProvider;
+use icu_provider::prelude::*;
+use std::convert::TryFrom;
+use std::path::PathBuf;
+
+mod patterns;
+mod skeletons;
+mod symbols;
 pub mod week_data;
+
+/// Common code for a data provider reading from CLDR JSON dates files.
+#[derive(Debug)]
+pub struct CommonDateProvider {
+    paths: Vec<(&'static str, &'static str, PathBuf)>,
+    data: FrozenBTreeMap<ResourceOptions, Box<cldr_serde::ca::Dates>>,
+}
+
+impl TryFrom<&dyn CldrPaths> for CommonDateProvider {
+    type Error = Error;
+    fn try_from(cldr_paths: &dyn CldrPaths) -> Result<Self, Self::Error> {
+        Ok(Self {
+            paths: cldr_paths.cldr_dates_all(),
+            data: FrozenBTreeMap::new(),
+        })
+    }
+}
+
+macro_rules! impl_resource_provider {
+    ($(($marker:ident, $expr:expr)),+) => {
+        $(
+            impl ResourceProvider<$marker> for CommonDateProvider {
+                fn load_resource(
+                    &self,
+                    req: &DataRequest,
+                ) -> Result<DataResponse<$marker>, DataError> {
+                    let langid = req
+                        .get_langid()
+                        .ok_or_else(|| DataErrorKind::NeedsLocale.with_req(<$marker>::KEY, req))?;
+                    let variant = req
+                        .options
+                        .variant
+                        .as_ref()
+                        .ok_or_else(|| DataErrorKind::NeedsVariant.with_req(<$marker>::KEY, req))?;
+
+                    let dates = self.data.get_or_insert(&req.options, || -> Result<_, DataError> {
```

Comment:
Thinking about this a bit more, `get_or_insert` feels doable with just a get and insert (and has the same perf), I think I'd prefer to remove the API, thoughts?

Reply:
@robertbastian try it out with elsa master and once you confirm it works I'll cut a new version

Reply:
That was my one method :(. Doing separate `get`s and `inserts` adds control flow to the caller which makes the code more complex. Option also has `get_or_insert`, presumably for this reason.

Reply:
I did update the code

Reply:
Yeah I saw, I'll get you your version bump in a bit